dbms/src/Storages/S3/FileCache.cpp

```diff
@@ -0,0 +1,563 @@
+        ostr << result.GetBody().rdbuf();
+        RUNTIME_CHECK_MSG(ostr.good(), "Write {} failed: {}", temp_fname, strerror(errno));
+        ostr.flush();
+    }
+    std::filesystem::rename(temp_fname, local_fname);
+    file_seg->setStatus(FileSegment::Status::Complete);
+    LOG_DEBUG(log, "Download s3_key={} to local={} size={} cost={}ms", s3_key, local_fname, content_length, sw.elapsedMilliseconds());
+}
+
+void FileCache::download(const String & s3_key, FileSegmentPtr & file_seg)
+{
+    try
+    {
+        downloadImpl(s3_key, file_seg);
+    }
+    catch (...)
+    {
+        tryLogCurrentException(log, fmt::format("Download s3_key={} failed", s3_key));
+    }
+
+    if (!file_seg->isReadyToRead())
+    {
+        bg_download_fail_count.fetch_add(1, std::memory_order_relaxed);
+        file_seg.reset();
+        remove(s3_key);
+    }
+    else
+    {
+        bg_download_succ_count.fetch_add(1, std::memory_order_relaxed);
+    }
+    bg_downloading_count.fetch_sub(1, std::memory_order_relaxed);
+    LOG_DEBUG(log, "downloading count {} => s3_key {} finished", bg_downloading_count.load(std::memory_order_relaxed), s3_key);
+}
+
+void FileCache::bgDownload(const String & s3_key, FileSegmentPtr & file_seg)
+{
+    bg_downloading_count.fetch_add(1, std::memory_order_relaxed);
+    LOG_DEBUG(log, "downloading count {} => s3_key {} start", bg_downloading_count.load(std::memory_order_relaxed), s3_key);
+    IOThreadPool::get().scheduleOrThrowOnError(
+        [this, s3_key = s3_key, file_seg = file_seg]() mutable {
+            download(s3_key, file_seg);
+        });
+}
+
+bool FileCache::isS3Filename(const String & fname)
+{
+    return S3::S3FilenameView::fromKey(fname).isValid();
+}
+
+String FileCache::toLocalFilename(const String & s3_key)
+{
+    return fmt::format("{}/{}", cache_dir, s3_key);
+}
+
+String FileCache::toS3Key(const String & local_fname)
+{
+    return local_fname.substr(cache_dir.size() + 1);
+}
+
+String FileCache::toTemporaryFilename(const String & fname)
+{
+    std::filesystem::path p(fname);
+    if (p.extension() != ".tmp")
+    {
+        return fmt::format("{}.tmp", fname);
+    }
+    else
+    {
+        return fname;
+    }
+}
+
+bool FileCache::isTemporaryFilename(const String & fname)
+{
+    std::filesystem::path p(fname);
+    return p.extension() == ".tmp";
+}
+
+void FileCache::prepareDir(const String & dir)
+{
+    std::filesystem::path p(dir);
+    RUNTIME_CHECK_MSG(!std::filesystem::exists(p) || std::filesystem::is_directory(p), "{} exists but not directory", dir);
+    if (!std::filesystem::exists(p))
+    {
+        std::filesystem::create_directories(p);
+    }
+}
+
+void FileCache::prepareParentDir(const String & local_fname)
+{
+    auto pos = local_fname.find_last_of('/');
+    RUNTIME_CHECK_MSG(pos != std::string::npos, "local filename {} is invalid", local_fname);
+    auto parent_dir = local_fname.substr(0, pos);
+    prepareDir(parent_dir);
+}
+
+void FileCache::restore()
```

Comment:
I think we need to use multiple threads to run this method. Or use background task to run the restore process.

Reply:
Let's do it later.